File: src/motor_driver_cpp_node/src/motor_driver.cpp
#include <wiringPi.h>
#include <iostream>
#include <cmath>

class MotorDriver {
public:
    // GPIO 引脚定义（BCM编号）
    const int PWMA = 18;
    const int AIN1 = 14;
    const int AIN2 = 15;
    const int PWMB = 19;
    const int BIN1 = 23;
    const int BIN2 = 24;

    const int AL = 16;
    const int BL = 20;
    const int AR = 26;
    const int BR = 21;

    const double MAX_SPEED = 1.0;

    volatile long left_ticks = 0;
    volatile long right_ticks = 0;

    MotorDriver() {
        wiringPiSetupGpio(); // 使用 BCM 编号

        pinMode(AIN1, OUTPUT);
        pinMode(AIN2, OUTPUT);
        pinMode(BIN1, OUTPUT);
        pinMode(BIN2, OUTPUT);
        pinMode(PWMA, PWM_OUTPUT);
        pinMode(PWMB, PWM_OUTPUT);

        // 初始化 PWM
        pwmSetMode(PWM_MODE_MS);
        pwmSetRange(1024);
        pwmSetClock(32);

        // 编码器输入引脚
        pinMode(AL, INPUT);
        pinMode(BL, INPUT);
        pinMode(AR, INPUT);
        pinMode(BR, INPUT);

        pullUpDnControl(AL, PUD_UP);
        pullUpDnControl(BL, PUD_UP);
        pullUpDnControl(AR, PUD_UP);
        pullUpDnControl(BR, PUD_UP);

        // 注册中断回调
        wiringPiISR(AL, INT_EDGE_RISING, &MotorDriver::left_callback_wrapper);
        wiringPiISR(AR, INT_EDGE_RISING, &MotorDriver::right_callback_wrapper);

        instance = this;
    }

    ~MotorDriver() {
        stop();
    }

    // 设置左电机速度（单位：m/s）
    void set_left_motor(double velocity) {
        double duty = std::min(std::abs(velocity) / MAX_SPEED, 1.0);
        int pwm_value = (int)(duty * 1023);

        if (velocity > 0) {
            digitalWrite(BIN1, LOW);
            digitalWrite(BIN2, HIGH);
        } else if (velocity < 0) {
            digitalWrite(BIN1, HIGH);
            digitalWrite(BIN2, LOW);
        } else {
            digitalWrite(BIN1, LOW);
            digitalWrite(BIN2, LOW);
        }
        pwmWrite(PWMB, pwm_value);
    }

    // 设置右电机速度
    void set_right_motor(double velocity) {
        double duty = std::min(std::abs(velocity) / MAX_SPEED, 1.0);
        int pwm_value = (int)(duty * 1023);

        if (velocity > 0) {
            digitalWrite(AIN1, LOW);
            digitalWrite(AIN2, HIGH);
        } else if (velocity < 0) {
            digitalWrite(AIN1, HIGH);
            digitalWrite(AIN2, LOW);
        } else {
            digitalWrite(AIN1, LOW);
            digitalWrite(AIN2, LOW);
        }
        pwmWrite(PWMA, pwm_value);
    }

    // 停止所有电机
    void stop() {
        pwmWrite(PWMA, 0);
        pwmWrite(PWMB, 0);
        digitalWrite(AIN1, LOW);
        digitalWrite(AIN2, LOW);
        digitalWrite(BIN1, LOW);
        digitalWrite(BIN2, LOW);
    }

    // 获取编码器计数
    void get_ticks(long &left, long &right) {
        left = left_ticks;
        right = right_ticks;
    }

private:
    // 静态回调函数包装（wiringPiISR 不支持成员函数）
    static void left_callback_wrapper() {
        if (instance) instance->callback_left();
    }
    static void right_callback_wrapper() {
        if (instance) instance->callback_right();
    }

    void callback_left() {
        if (digitalRead(AL)) {
            if (!digitalRead(BL))
                left_ticks--;
            else
                left_ticks++;
        }
    }

    void callback_right() {
        if (digitalRead(AR)) {
            if (!digitalRead(BR))
                right_ticks++;
            else
                right_ticks--;
        }
    }

    static MotorDriver* instance;
};

File: src/motor_driver/motor_driver/motor_driver.py
import RPi.GPIO as GPIO
import time
from motor_driver.pid_pwm import PIDPWM
class MotorDriver():
    def __init__(self):
        self.PWMA = 18
        self.AIN1 = 14
        self.AIN2 = 15
        self.PWMB = 19
        self.BIN1 = 23
        self.BIN2 = 24
        self.MAX_SPEED = 1.0
        self.min_duty = 5
        self.left_ticks = 0
        self.right_ticks = 0

        self.E2A = 16
        self.E2B = 20
        self.E1A = 26
        self.E1B = 21

        GPIO.setmode(GPIO.BCM)
        GPIO.setup([self.AIN1, self.AIN2, self.BIN1, self.BIN2], GPIO.OUT)
        GPIO.setup([self.PWMA, self.PWMB], GPIO.OUT)
        self.pwma = GPIO.PWM(self.PWMA, 1000)
        self.pwmb = GPIO.PWM(self.PWMB, 1000)
        self.pwma.start(0)
        self.pwmb.start(0)
        self.pid_pwma = PIDPWM(self.pwma)
        self.pid_pwmb = PIDPWM(self.pwmb)
        self.pid_pwma.start()
        self.pid_pwmb.start()
        GPIO.setup(self.E2B, GPIO.IN, pull_up_down=GPIO.PUD_UP)
        GPIO.setup(self.E2A, GPIO.IN, pull_up_down=GPIO.PUD_UP)
        GPIO.setup(self.E1B, GPIO.IN, pull_up_down=GPIO.PUD_UP)
        GPIO.setup(self.E1A, GPIO.IN, pull_up_down=GPIO.PUD_UP)

        GPIO.add_event_detect(self.E2A, GPIO.RISING, callback=self.callback_left)
        GPIO.add_event_detect(self.E1A, GPIO.RISING, callback=self.callback_right)

    def set_left_motor(self, velocity):
        duty = 0.0
        if velocity > self.MAX_SPEED:
            duty = 100.0
        else:
            duty = abs(velocity) / self.MAX_SPEED * 100.0
        if abs(velocity) > 0:
            duty = max(self.min_duty, duty)
        else: 
            duty = 0
        if velocity > 0:
            GPIO.output(self.BIN1, GPIO.LOW)
            GPIO.output(self.BIN2, GPIO.HIGH)
        elif velocity < 0:
            GPIO.output(self.BIN1, GPIO.HIGH)
            GPIO.output(self.BIN2, GPIO.LOW)
        else:
            GPIO.output(self.BIN1, GPIO.LOW)
            GPIO.output(self.BIN2, GPIO.LOW)

        self.pid_pwmb.set_target(duty)
        return duty
    
    def set_right_motor(self, velocity):
        duty = 0.0
        if velocity > self.MAX_SPEED:
            duty = 100.0
        else:
            duty = abs(velocity) / self.MAX_SPEED * 100.0
        if abs(velocity) > 0:
            duty = max(self.min_duty, duty)
        else: 
            duty = 0
        if velocity > 0:
            GPIO.output(self.AIN1, GPIO.LOW)
            GPIO.output(self.AIN2, GPIO.HIGH)
        elif velocity < 0:
            GPIO.output(self.AIN1, GPIO.HIGH)
            GPIO.output(self.AIN2, GPIO.LOW)
        else:
            GPIO.output(self.AIN1, GPIO.LOW)
            GPIO.output(self.AIN2, GPIO.LOW)
        self.pid_pwma.set_target(duty)
        return duty
    
    def stop(self):
        self.pid_pwma.stop()
        self.pid_pwmb.stop()
        GPIO.cleanup()

    def get_ticks(self):
        return self.left_ticks, self.right_ticks

    def callback_left(self, channel):
        if GPIO.input(self.E2A):
            if not GPIO.input(self.E2B):
                self.left_ticks -= 1
            elif GPIO.input(self.E2B):
                self.left_ticks += 1

    def callback_right(self, channel):
        if GPIO.input(self.E1A):
            if not GPIO.input(self.E1B):
                self.right_ticks += 1
            elif GPIO.input(self.E1B):
                self.right_ticks -= 1

    